src/libImaging/TiffDecode.c

```diff
@@ -246,77 +285,164 @@ _decodeStripYCbCr(Imaging im, ImagingCodecState state, TIFF *tiff) {
     /* malloc check above */
     new_data = realloc(state->buffer, state->bytes);
     if (!new_data) {
         state->errcode = IMAGING_CODEC_MEMORY;
-        goto decodeycbcr_err;
+        goto decodergba_err;
     }
 
     state->buffer = new_data;
 
-    for (; state->y < state->ysize; state->y += rows_per_strip) {
+    for (; state->y < state->ysize; state->y += rows_per_block) {
         img.row_offset = state->y;
-        rows_to_read = min(rows_per_strip, img.height - state->y);
+        rows_to_read = min(rows_per_block, img.height - state->y);
 
         if (TIFFRGBAImageGet(&img, (UINT32 *)state->buffer, img.width, rows_to_read) ==
             -1) {
             TRACE(("Decode Error, y: %d\n", state->y));
             state->errcode = IMAGING_CODEC_BROKEN;
-            goto decodeycbcr_err;
+            goto decodergba_err;
         }
 
+#if WORDS_BIGENDIAN
+        TIFFSwabArrayOfLong((UINT32 *)state->buffer, img.width * rows_to_read);
+#endif
+
         TRACE(("Decoded strip for row %d \n", state->y));
 
         // iterate over each row in the strip and stuff data into image
-        for (strip_row = 0;
-             strip_row < min((INT32)rows_per_strip, state->ysize - state->y);
-             strip_row++) {
-            TRACE(("Writing data into line %d ; \n", state->y + strip_row));
+        for (current_row = 0;
+             current_row < min((INT32)rows_per_block, state->ysize - state->y);
+             current_row++) {
+            TRACE(("Writing data into line %d ; \n", state->y + current_row));
 
-            // UINT8 * bbb = state->buffer + strip_row * (state->bytes /
-            // rows_per_strip); TRACE(("chars: %x %x %x %x\n", ((UINT8 *)bbb)[0],
+            // UINT8 * bbb = state->buffer + current_row * (state->bytes /
+            // rows_per_block); TRACE(("chars: %x %x %x %x\n", ((UINT8 *)bbb)[0],
             // ((UINT8 *)bbb)[1], ((UINT8 *)bbb)[2], ((UINT8 *)bbb)[3]));
 
             state->shuffle(
-                (UINT8 *)im->image[state->y + state->yoff + strip_row] +
+                (UINT8 *)im->image[state->y + state->yoff + current_row] +
                     state->xoff * im->pixelsize,
-                state->buffer + strip_row * row_byte_size,
+                state->buffer + current_row * row_byte_size,
                 state->xsize);
         }
     }
 
-decodeycbcr_err:
+decodergba_err:
     TIFFRGBAImageEnd(&img);
     if (state->errcode != 0) {
         return -1;
     }
     return 0;
 }
 
+int
+_decodeTile(Imaging im, ImagingCodecState state, TIFF *tiff, int planes, ImagingShuffler *unpackers) {
+    INT32 x, y, tile_y;
+    UINT32 tile_width, tile_length, current_tile_length, current_line,
+        current_tile_width, row_byte_size;
+    UINT8 *new_data;
+
+    TIFFGetField(tiff, TIFFTAG_TILEWIDTH, &tile_width);
+    TIFFGetField(tiff, TIFFTAG_TILELENGTH, &tile_length);
+
+    /* overflow check for row_byte_size calculation */
+    if ((UINT32)INT_MAX / state->bits < tile_width) {
```

Comment:
There's a UINT32_MAX, this will be 2^31 iirc. Are you aiming for 2^31 or 2^32?

Reply:
That's a good point.
I am going to check on this during this week.
I didn't change that check, just moved it into the function.

Reply:
Looked into that further and discovered that `state->bytes` is plain `int`, so can't hold uint32 values. Additionally, image dimensions (`im->xsize`, `im->ysize`) and dimensions on the state (`state->x`, `state->y`, `state->xsize`, ... ) are also `int`.
I decided to trust LibTiff's `TIFFStripSize`/`TIFFTileSize` to simplify calculations of row size in pixels and in bytes, since subsampled YCbCr images are handled in another case. These functions return `tsize_t` value which is [same as `size_t` but signed](https://gitlab.com/libtiff/libtiff/-/blob/master/libtiff/tiffio.h#L40-76).
Anyway, making sure that tile/strip dimensions and buffer bytes count are within INT_MAX seems safest.

However, switching to use `TIFFStripSize`/`TIFFTileSize` I saw crash tests crash, so had to put back check that LibTiff would return expected amount of data per tile/strip - hence 2 commits